src/languages/php.js

```diff
@@ -18,6 +18,28 @@ export default function(hljs) {
       // Perl at all like $ident$, @ident@, etc.
       `(?![A-Za-z0-9])(?![$])`
   };
+  const CONSTANT = {
+    className: "variable",
+    relevance: 0,
+    variants: [
+      {
+        begin: /const\s*/,
+        end: /\w+/,
+        excludeBegin: true
+      },
+      {
+        begin: /::(?!class\b)/,
+        end: /\w+/,
+        excludeBegin: true
+      },
+      {
+        className: "keyword",
+        begin: /::/,
+        end: /class/,
+        excludeBegin: true
+      },
```

Comment:
Not sure what this rule is try too accomplish?

Reply:
[class](https://wiki.php.net/rfc/class_name_literal_on_object) is a special case of the constant, it's also a reserved keyword.

Reply:
Ok, I see what you're doing nwo... I'd use a multi-match here also but I'd consider going even wider and getting the class name as well.  Doesn't `::class` imply we have a class?

```
[DateTimeImmutable][::][class]

DateTimeImmutable => title.class
class => keyword
```

Reply:
> Doesn't ::class imply we have a class?

Nope. It can be also the keyword `self` or `static` that is an alias for the class in which they are called or an object (variable).



    public function getClassFromSelf(): string {
        return self::class;
    }
    public static function getClassFromStaic(): string {
        return static::class;
    }
    public function getClass(): string {
        $date = new DateTimeImmutable();
        return date::class;
    }

Reply:
Does the uppercase _by convention_ generally mean we have a class?  ie `Car::class`?  We do this for other grammars if it's a strong convention even though nothing requires classes to be uppercase in say Javascript.

Reply:
> Does the uppercase by convention generally mean we have a class? 



It's a good practice, so I implemented it that way.

Reply:
About the `self` and `static` I didn't notice earlier that they are marked as `hljs-built_in` and not keywords.
My current implementation changed this behavior inside the function. Now they are marked as a class. Tomorrow I will try to find this regression.

Reply:
I wonder if we couldn't deal with this by just exhausting all the non-class possibilities... ie:

- `self::class`
- `static::class`
- `$[variable name]::class`
- [whatever]::class` (must be a class, no other options)

Is that perhaps doable?

Reply:
There are also per-rule JS callbacks that could manually evaluate the names and make a decision, but we don't encourage their usage (speed)... so I'd likely first want to:

- see if there is some way to make this sorta-work case-insensative
- see if the entire grammar could work as case-sensative

And only if both those fail look into using call-backs to do the class or not-class classification.

Reply:
Also now that I'm thinking about it hard  `static::class` the class here does not look like a keyword... ie, the context it's not being used as a keyword... perhaps `variable.language` or `variable.constant` would be more appropriate... see the CSS class docs. 

IE, in Ruby:

```rb
# here class is a keyword
class Dog
end

# here it's just a built-in property of Dog that returns Dog's meta-class
a = Dog.class
```

The first should be `title.class`, the second `property`.

Reply:
@joshgoebel Changed to `variable.constant`